### PR TITLE
Feature/swrl rule with functional

### PR DESCRIPTION
**Pull request by althonos**
This is based on #84, I still need to fix a few bugs and implement the serializer.

### CODE REVIEW

**Comment by phillord-ncl** [comment]
Looks good so far.

Can you PR it against the feature/swrl_rule_with_function branch rather than devel? While I was playing with your code, I've started to fix a few other things (basically, updating horned-bin to work with ofn); I can add those after you've finished.

Then we can merge to everything to devel and take stock of what, if anything, needs to be done before 1.0.

**Comment by althonos** [comment]
I'm done implementing the writer as well.

The one strange thing is that the order or the body and head atoms in the functional files seems to be always reversed from the XML files; I made the parser and the serializer reverse the atom vectors when reading/writing but that seems a bit strange, I suppose that's an implementation detail from `bubo`.

**Review by filippodebortoli — APPROVED** [review]

**Comment by phillord** [comment]
Not sure I understand the "reversed" bit.

```
 <DLSafeRule>
        <Body>
            <ClassAtom>
                <Class IRI="#A"/>
                <Variable IRI="#x"/>
            </ClassAtom>
        </Body>
        <Head>
            <ClassAtom>
                <Class IRI="#B"/>
                <Variable IRI="#x"/>
            </ClassAtom>
        </Head>
    </DLSafeRule>
```

in XML while functional:

```
DLSafeRule(Body(ClassAtom(o:A Variable(o:x)))Head(ClassAtom(o:B Variable(o:x))))
)
```

If I had been writing the spec, I would have done "head" then "body", if that is what you mean.

File: src/io/ofn/reader/from_pair.rs [hunk 1]
```diff
@@ -876,7 +984,8 @@ macro_rules! impl_ontology {
                     let inner = pair.into_inner().next().unwrap();
                     match inner.as_rule() {
                         // FIXME: SWRL rules are not supported for now
-                        Rule::Rule | Rule::DGAxiom => (),
+                        // Rule::DGAxiom => unimplemented!(),
```
Comment (phillord) [inline src/io/ofn/reader/from_pair.rs:987]: Can you remove the comments!

File: src/io/ofn/writer/as_functional.rs [hunk 2]
```diff
@@ -434,14 +440,82 @@ impl<A: ForIRI> AsFunctional<A> for AnnotationValue<A> {}
...
     }
 }
 
 impl<A: ForIRI> AsFunctional<A> for AnonymousIndividual<A> {}
 
 // ---------------------------------------------------------------------------
 
+impl<'a, A: ForIRI> Display for Functional<'a, Atom<A>, A> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        use Atom::*;
```
Comment (phillord) [inline src/io/ofn/writer/as_functional.rs:453]: I didn't know you could do that! Learn something every day.

**Comment by althonos** [comment]
> Not sure I understand the "reversed" bit.

This can be seen in the `swrl_two_variables` files. In XML, you have in the `A1` then `A` in the body, and `B1` then `B` in the head:
```xml
    <DLSafeRule>
        <Body>
            <ClassAtom>
                <Class IRI="#A1"/>
                <Variable IRI="#x"/>
            </ClassAtom>
            <ClassAtom>
                <Class IRI="#A"/>
                <Variable IRI="#x"/>
            </ClassAtom>
        </Body>
        <Head>
            <ClassAtom>
                <Class IRI="#B1"/>
                <Variable IRI="#x"/>
            </ClassAtom>
            <ClassAtom>
                <Class IRI="#B"/>
                <Variable IRI="#x"/>
            </ClassAtom>
        </Head>
    </DLSafeRule>
```
 whereas in functional you have `A` then `A1` in body, and `B` then `B1` in head:
```
DLSafeRule(
    Body(ClassAtom(o:A Variable(o:x)) ClassAtom(o:A1 Variable(o:x)))
    Head(ClassAtom(o:B Variable(o:x)) ClassAtom(o:B1 Variable(o:x)))
)
```

**Comment by phillord-ncl** [comment]
Ah, okay, I see the point. Both head and body are sets in the OWL API, so the ordering isn't defined. But I would have expected it to be consistent because the ordering of a `Set` in Java usually is and the XML and functional files are produced from exactly the same objects in memory.

We could have a poke in the OWL API to see why this happens. Be curious to see what robot does also.

Does this cause problems in testing? I had to add a normalisation step to the OWX->RDF comparison to cope with sets.

**Comment by althonos** [comment]
> Both head and body are sets in the OWL API, so the ordering isn't defined. 

I assumed that was the case.

> Does this cause problems in testing?

Not really, since this only appears in one test case, and as I said I just had to reverse the vectors, but if those are sets this may not be consistent if more test cases are added. Maybe would be a good idea to check the OWL API, perhaps this should be dumped in lexicographic order instead?

**Comment by phillord** [comment]
Indeed. Perhaps @ignazio1977 knows how the ordering works.

**Comment by ignazio1977** [comment]
Depends on the exact version in use. It used to sort the sets, was changed to a linked set on user request, so order is the order the terms were first created in.

**Comment by phillord** [comment]
Confusing. As far as I can see the relevant code is in SimpleRenderer for OFN.

```
    @Override
    public void visit(SWRLRule rule) {
        sb.append("DLSafeRule(");
        writeAnnotations(rule);
        sb.append("Body(");
        render(rule.getBody(), false);
        sb.append(") Head(");
        render(rule.getHead(), false);
        sb.append("))");
    }
```

the "false" in the render call means, "don't sort". While for OWX we have in     OWLXMLObjectRenderer

```
    @Override
    public void visit(SWRLRule rule) {
        writer.writeStartElement(DL_SAFE_RULE);
        writeAnnotations(rule);
        writer.writeStartElement(BODY);
        for (SWRLAtom atom : rule.getBody()) {
            atom.accept(this);
        }
        writer.writeEndElement();
        writer.writeStartElement(HEAD);
        for (SWRLAtom atom : rule.getHead()) {
            atom.accept(this);
        }
        writer.writeEndElement();
        writer.writeEndElement();
    }
```


So, both use the iteration order of the set. This is for version4 which bubo uses. For version5 it is the stream order.

I have no idea why the two are coming out differently. It also means that we are dependent `Set` implementation ordering at best. So I don't think we can do anything to be consistent with the OWL API here that is guaranteed, and the OWL API doesn't give us guarantees.

**Comment by phillord** [comment]
> Depends on the exact version in use. It used to sort the sets, was changed to a linked set on user request, so order is the order the terms were first created in.

Ah, okay. But why it is inconsistent between the two formats? Or have I been unlucky in generated XML with one version (4.5.26 according to the comments) and OFN with another (version unknown as there are no comments).

**Comment by ignazio1977** [comment]
It shouldn't be inconsistent.

The order can be decided by the renderer, but the default is the order coming from the implementation of getHead/getBody, which are backed with lists, if memory serves.

Behaviour has changed in patch versions as well. It's meant to be the same in the latest 4.x and 5.x.

**Comment by phillord** [comment]
And,yet, it appears to be. I've just regenerated the files for swrl_two_variables.clj and still get the same thing -- the XML output gives the body axioms in the opposite order to the functional syntax.

RDF and ttl output seem both to be in the same order as XML.

I've tested with using Robot to convert from RDF to XML and OFN. Same thing.

**Comment by phillord** [comment]
I have tried lots of experiments here. My conclusion is that I am very confused.

This ontology

```
DLSafeRule(Body(ClassAtom(o:A Variable(o:x)) ClassAtom(o:A1 Variable(o:x)))Head(ClassAtom(o:B Variable(o:x)) ClassAtom(o:B1 Variable(o:x))))
```

renders as this 

```
            <ClassAtom>
                <Class IRI="#A1"/>
                <Variable IRI="#x"/>
            </ClassAtom>
            <ClassAtom>
                <Class IRI="#A"/>
                <Variable IRI="#x"/>
            </ClassAtom>
```

So the class atoms are the other way around.

I have tried with it out with a variety of ontologies with different class axioms. So this:

```
DLSafeRule(Body(ClassAtom(o:Z1 Variable(o:x)) ClassAtom(o:Y2 Variable(o:x)) ClassAtom(o:X3 Variable(o:x)) ClassAtom(o:W4 Variable(o:x)))Head(ClassAtom(o:B Variable(o:x)) ClassAtom(o:B1 Variable(o:x))))
```

renders as 

```
            <ClassAtom>
                <Class IRI="#Z1"/>
                <Variable IRI="#x"/>
            </ClassAtom>
            <ClassAtom>
                <Class IRI="#Y2"/>
                <Variable IRI="#x"/>
            </ClassAtom>
            <ClassAtom>
                <Class IRI="#X3"/>
                <Variable IRI="#x"/>
            </ClassAtom>
            <ClassAtom>
                <Class IRI="#W4"/>
                <Variable IRI="#x"/>
            </ClassAtom>

```

with the atoms in the same order. 

So I have no clear idea what could be happening. The only thing that comes to mind is that `Set` instance I am passing into the OWL API might be used directly by OWL and that this set might have a non-stable iteration order. In which case,  it is probably a Clojure oddity rather than OWL API.

**Comment by phillord** [comment]
And that appears to be it. If I pass a `java.util.TreeSet` rather than a `clojure.lang.PersistantSet` into  `getSWRLRule` then the problem goes away and we get a stable order between the two. How entirely unexpected!

**Comment by phillord** [comment]
I guess as we have a rough idea of the problem, but we don't know if it is going to turn into an actual problem with tests, we should merge this in now. @althonos are you happy with this?

**Comment by althonos** [comment]
@phillord: If you start having more problems with testing later we can always sort the body/head in both parsers, but since for now it only causes a problem with one example (where we're lucky I guess) we can probably merge this :+1:

**Comment by phillord** [comment]
All done!

**Pull request merged**